lib/pos_mempool.go

```diff
@@ -542,16 +585,87 @@ func (mp *PosMempool) updateTransactionValidatedStatus(txnHash *BlockHash, valid
-	userPk := NewPublicKey(txn.Tx.PublicKey)
+func (mp *PosMempool) checkNonceTracker(txn *MempoolTx, userPk *PublicKey) (*MempoolTx, error) {
 
 	// Check the nonceTracker to see if this transaction is meant to replace an existing one.
 	existingTxn := mp.nonceTracker.GetTxnByPublicKeyNonce(*userPk, *txn.Tx.TxnNonce)
 	if existingTxn != nil && existingTxn.FeePerKB > txn.FeePerKB {
-		return errors.Wrapf(MempoolFailedReplaceByHigherFee, "PosMempool.AddTransaction: Problem replacing transaction "+
+		return nil, errors.Wrapf(MempoolFailedReplaceByHigherFee, "PosMempool.AddTransaction: Problem replacing transaction "+
 			"by higher fee failed. New transaction has lower fee.")
 	}
 
+	// TODO: is it okay to allow if the incoming tx is an inner atomic txn?
+	if existingTxn != nil && existingTxn.Tx.IsAtomicTxnsInnerTxn() {
+		return nil, errors.Wrapf(MempoolFailedReplaceByHigherFee, "PosMempool.AddTransaction: Cannot replace txn that is"+
+			"an inner atomic txn.")
+	}
+	return existingTxn, nil
+}
+
+func (mp *PosMempool) removeNonces(txns []*MsgDeSoTxn) {
+	for _, txn := range txns {
+		userPk := NewPublicKey(txn.PublicKey)
+		mp.nonceTracker.RemoveTxnByPublicKeyNonce(*userPk, *txn.TxnNonce)
+	}
+}
+
+func (mp *PosMempool) persistMempoolAddEvent(txn *MempoolTx, persistToDb bool) {
+	// Emit an event for the newly added transaction.
+	if persistToDb && !mp.inMemoryOnly {
+		event := &MempoolEvent{
+			Txn:  txn,
+			Type: MempoolEventAdd,
+		}
+		mp.persister.EnqueueEvent(event)
+	}
+}
+
+func (mp *PosMempool) addTransactionNoLock(txn *MempoolTx, persistToDb bool) error {
+	userPk := NewPublicKey(txn.Tx.PublicKey)
+
+	// Special handling for atomic txns. For atomic txns, the mempool will ignore the nonce for the wrapper txn
+	// and only track nonces for the inner txns. Additionally, only the wrapper txn will be added to the transaction
+	// register.
+	if txn.Tx.TxnMeta.GetTxnType() == TxnTypeAtomicTxnsWrapper {
+		// If the txn is an atomic txn, we need to add each txn individually.
+		atomicTxnsWrapper, ok := txn.Tx.TxnMeta.(*AtomicTxnsWrapperMetadata)
+		if !ok {
+			return fmt.Errorf(
+				"PosMempool.AddTransaction: Problem adding atomic txn - casting metadata failed")
+		}
+		var innerMempoolTxs []*MempoolTx
+		for _, innerTxn := range atomicTxnsWrapper.Txns {
+			newInnerMempoolTx, err := NewMempoolTx(innerTxn, txn.Added, uint64(txn.Height))
+			if err != nil {
+				return errors.Wrapf(err, "PosMempool.AddTransaction: Problem creating MempoolTx from inner atomic txn")
+			}
+			innerMempoolTxs = append(innerMempoolTxs, newInnerMempoolTx)
+		}
+		for _, innerMempoolTx := range innerMempoolTxs {
+			innerUserPk := NewPublicKey(innerMempoolTx.Tx.PublicKey)
+			if _, err := mp.checkNonceTracker(innerMempoolTx, innerUserPk); err != nil {
+				// if we hit an error, we need to remove all the nonces from the nonce tracker.
+				mp.removeNonces(atomicTxnsWrapper.Txns)
```

Comment:
Wouldn't you want to just remove the nonces you added, and not all nonces from the atomicTxnsWrapper? There's probably some edge case where an inner txn in the atmoicTxnsWrapper reuses a nonce from another txn that was independently added to the mempool. You don't want any side effects that affect that txn